# Project 5 -- Kennedy Zapalac

**Piazza Help:**
- Question 1: asked if it was ok to search the whole line for the text, and Dr. Ward said yes

**Websites:**
- I used [this website](https://www.geeksforgeeks.org/awk-command-unixlinux-examples/) to learn more about the functionality of `awk` for question 5

In [ ]:
%%bash
# awk format just for my reference
awk -F";" 'BEGIN{ } { } END{ }'

# The -F is for the delimeter that occurs between pieces of data in your file

# The BEGIN portion of an awk program runs at the ver start before we process the file, for instance, in you want to print a header ad the top of the out, then put it inot the BEGIN section of an awk program

# # The main portion of an awk program runs once per line of the data
# This is definitely the most important part of an awk program!

# The END portion of an awk porgram runs at the very end of the processing of the file just one time for instance, if you to print a summary of what you foudn in the data in some way, it often goes in the END portion of an awk program

# The BEGIN and END portions are actually optional so many awk programs just look like this:

awk -F";" '{ the stuff in here runs one time per line of your data set }'

# some awk prgrams look like this, without any BEGIN but they do have an END:

awk -F";" '{ } END{ }'

## Question 1
A. What is the total cost of purchases with `WHISKIES` in the title?

B. What is the total cost of all purchases from `CEDAR RAPIDS` (not just `WHISKIES`; consider all purchases)

In [13]:
%%bash
# question 1a
echo "Total cost of purchases with WHISKIES in the line"
# searching for whiskies in the whole line and adding state bottle price to myprices if the line contains that
# was told it's ok to search the whole line
# repeat for each line
# finally print the total cost of purchases with whiskies in the line, represented by myprices
cat /anvil/projects/tdm/data/iowa_liquor_sales/iowa_liquor_sales_cleaner.txt | awk -F";" '{if ($0 ~ /WHISKIES/) {myprices += $19}} END{print myprices}'


Total cost of purchases with WHISKIES in the line
8.59911e+07


In [14]:
%%bash
# question 1b
echo "Total cost of purchases from CEDAR RAPIDS"
# searching for whiskies in the whole line and adding state bottle price to myprices if the line contains that
# repeat for each line
# finally print the total cost of purchases with cedar rapids in the line, represented by myprices
cat /anvil/projects/tdm/data/iowa_liquor_sales/iowa_liquor_sales_cleaner.txt | awk -F";" '{if ($0 ~ /CEDAR RAPIDS/) {myprices += $19}} END{print myprices}'


Total cost of purchases from CEDAR RAPIDS
1.7924e+07


The total cost of purchases with `WHISKIES` in the title is 85,991,100 dollars and the total cost of purchases from `CEDAR RAPIDS` is 17,924,000 dollars

## Question 2
A. What `Store Name` had the largest number of purchases (not the largest total cost, but the largest number of purchases; please consider each line to be 1 purchase)

B. Using the `Store Name` identified Question 2A, what was the total cost of all purchases from this `Store Name`?

In [21]:
%%bash
# question 2a
# extract ONLY the field with the store names
# sort the resulting data so that the same store names are next to each other
# then use uniq -c to see how many values of each store name there are
# and sort this data using sort -n (to sort the data numerically)
# and finally take the tail of the results
cat /anvil/projects/tdm/data/iowa_liquor_sales/iowa_liquor_sales_cleaner.txt | cut -d";" -f4 | sort | uniq -c | sort -n | tail -n1

 223447 HY-VEE #3 / BDI / DES MOINES


In [5]:
%%bash
# question 2b
# pipe the output from cat to awk
# ; is the delimeter
# if field 4, the store name, contains "HY-VEE #3 / BDI / DES MOINES", then pull field 19, the state bottle cost, and sum the cost of all purchases from this store
# had to escape the forward slashes with backward slashes so awk didn't think the pattern to match ended
cat /anvil/projects/tdm/data/iowa_liquor_sales/iowa_liquor_sales_cleaner.txt | awk -F";" '{if ($4 ~ /HY-VEE #3 \/ BDI \/ DES MOINES/) {myprices += $19}} END{print myprices}'

2.88549e+06


The `STORE NAME` that had the largest numbers of purchases was "HY-VEE #3 / BDI / DES MOINES." The total cost of all purchases from this store was 2,885,490 dollars.

## Question 3
A. Please compute the total volume (in liters) of all purchases sold in the file `iowa_liquor_sales_cleaner.txt`

B. Please compute the total volume (in liters) of `VODKA 80 PROOF` sold in the file `iowa_liquor_sales_cleaner.txt`

In [10]:
%%bash
# Question 3a
# piping the ouput of cat to awk
# adding the value in field 23 to myvolumes for each line and printing the final sum
# field 23 = volume sold (liters)
cat /anvil/projects/tdm/data/iowa_liquor_sales/iowa_liquor_sales_cleaner.txt | awk -F";" '{myvolumes += $23} END{print myvolumes}'

2.47828e+08


In [11]:
%%bash
# Question 3b
# piping the output of cat to awk
# adding the value in field 23 to myvolumes only if the line contains `VODKA 80 PROOF`
# field 23 = volume sold (liters)
# repeat for each line
# print the final sum
cat /anvil/projects/tdm/data/iowa_liquor_sales/iowa_liquor_sales_cleaner.txt | awk -F";" '{if ($0 ~ /VODKA 80 PROOF/) {myvolumes += $23}} END{print myvolumes}'

1.78615e+07


The total volumne of all purchases sold in the file is 247,828,000 liters. The total volume of `VODKA 80 PROOF` sold in the file is 17,861,500 liters.

## Question 4
A. When looking at which location has the largest numbers of purchases, if we use the address (instead of the store name), we should include the `Address`, `City`, and `Zip Code`. Using these three variables (together), what location has the largest number of purchases?

B. Does your answer to Question 4A agree with your answer to Question 2A? How do you know? (Please explain why, and/or use some analysis to justify your answer.)

In [15]:
%%bash
# question 4a
# pipe the output of cat to cut
# extract the address, city, and zip code fields
# sort the resulting data so that the same store names are next to each other
# then use uniq -c to see how many values of each store name there are
# and sort this data using sort -n (to sort the data numerically)
# and finally take the tail of the results
cat /anvil/projects/tdm/data/iowa_liquor_sales/iowa_liquor_sales_cleaner.txt | cut -d";" -f5,6,7 | sort | uniq -c | sort -n | tail -n1

# alternate awk method
cat /anvil/projects/tdm/data/iowa_liquor_sales/iowa_liquor_sales_cleaner.txt | awk -F";" '{print $5, $6, $7}' | sort | uniq -c | sort -n | tail -n1

 223447 3221 SE 14TH ST;DES MOINES;50320
 223447 3221 SE 14TH ST DES MOINES 50320


3221 SE 14th St, Des Moines, 50320 has the largest number of purchases. This location occurs 223,447 times in our datasaet, meaning that there were 223,447 purchases at this location. The amount of purchases at this location agrees with the amount of purchases at the `STORE NAME` that had the largest number of purchases, which was 223,447  purchases at HY-VEE #3 / BDI / DES MOINES. The reason these 2 numbers agree is because each store name only has one location. So the store with the largest number of purchases is the same as the location with the largest number of purchases.

## Question 5
A. `awk` is powerful, and this liquor dataset is pretty interesting! We haven’t covered everything `awk` (and we won’t). Look at the dataset and ask yourself an interesting question about the data. Use `awk` to solve your problem (or, at least, get you closer to answering the question). Optionally: You can explore various stackoverflow questions about `awk` and `awk` guides online. Try to incorporate an `awk` function you haven’t used, or a awk trick you haven’t seen. While this last part is not required, it is highly encouraged and can be a fun way to learn something new.

Please be sure to put a brief explanation about your work in Question 5 using awk to study something interesting that YOU FOUND in the data in Question 5.

In [5]:
%%bash
# Question 5
# calculating the total volume in ml using the liters field
# just taking it a step further than question 3a
# Question 3a
# piping the ouput of cat to awk
# adding the value in field 23 to myvolumes for each line and printing the final sum divided by 1000
# field 23 = volume sold (liters)
echo "total volume sold in ml"
cat /anvil/projects/tdm/data/iowa_liquor_sales/iowa_liquor_sales_cleaner.txt | awk -F";" '{myvolumes += $23} END{print myvolumes/1000}'

total volume sold in ml
247828


I calculated the total volume sold in ml using the liters field because `awk` can also perform arithmetic operations! All I had to do was divide my outut in question 3a by 100. The total volume of all purchases sold is 247,828 ml. I looked at [this website](https://www.geeksforgeeks.org/awk-command-unixlinux-examples/) to find new ways I could use `awk`.

## Pledge

By submitting this work I hereby pledge that this is my own, personal work. I've acknowledged in the designated place at the top of this file all sources that I used to complete said work, including but not limited to: online resources, books, and electronic communications. I've noted all collaboration with fellow students and/or TA's. I did not copy or plagiarize another's work.

> As a Boilermaker pursuing academic excellence, I pledge to be honest and true in all that I do. Accountable together – We are Purdue.